oak_runtime/src/node/http/client.rs

```diff
@@ -0,0 +1,327 @@
+    }
+    !empty_root_ca && (config.scheme != UriScheme::Https as i32 || !config.authority.is_empty())
+}
+
+impl HttpClientNode {
+    /// Creates a new `HttpClientNode` that can either handle HTTP requests or HTTPS requests. To
+    /// handle HTTPS requests a non-empty authority in the `config` is required. This configuration
+    /// must also match the label with which this node is being created. If the node is configured
+    /// to handle URIs with the `HTTP` scheme, then it should be labelled public. In this case,
+    /// the authority could be empty, meaning that the node can handle all HTTP requests. In
+    /// addition, a non-empty `root_ca` is always required.
+    pub fn new(node_name: &str, config: HttpClientConfiguration, root_ca: Vec<u8>) -> Option<Self> {
+        if !validate_config(&config, root_ca.is_empty()) {
+            return None;
+        }
+        let http_client = create_client(root_ca).ok()?;
+        let node_privilege = get_privilege(config.authority.clone());
+        let authority = Some(config.authority).filter(|s| !s.is_empty());
+        Some(HttpClientNode {
+            node_name: node_name.to_string(),
+            http_client,
+            authority,
+            node_privilege,
+        })
+    }
+
+    /// Main loop that handles HTTP invocations from the `handle`, sends HTTP requests to an
+    /// external HTTP service and writes HTTP responses back to the invocation channel.
+    async fn handle_loop(&mut self, runtime: RuntimeProxy, handle: Handle) -> Result<(), OakError> {
+        // Create a [`Receiver`] to receive the HTTP request.
+        let receiver = Receiver::<Invocation>::new(ReadHandle { handle });
+        loop {
+            debug!("Waiting for HTTP invocation");
+            // Read an HTTP invocation from the [`Receiver`].
+            let invocation = receiver.receive(&runtime).map_err(|error| {
+                match error {
+                    OakError::OakStatus(OakStatus::ErrTerminated) => {
+                        debug!("HTTP client node is terminating.")
+                    }
+                    OakError::OakStatus(OakStatus::ErrChannelClosed) => {
+                        info!("HTTP invocation channel closed")
+                    }
+                    _ => error!("Couldn't receive the invocation: {:?}", error),
+                }
+                error
+            })?;
+
+            self.process_invocation(&runtime, &invocation).await;
+            debug!("Invocation processing finished");
+            invocation.close(&runtime);
+        }
+    }
+
+    /// Process an HTTP invocation for an external HTTP service.
+    async fn process_invocation(&mut self, runtime: &RuntimeProxy, invocation: &Invocation) {
+        // TODO(#1717): Update HTTP client metrics
+
+        // A closure to ease sending and logging the errors
+        let send_error = |err, code, msg| {
+            warn!("{}: {}", msg, err);
+            invocation.send_error(code, msg, runtime);
```

Comment:
Would it make sense to make this method return `Result` or something, and have the logging and sending the error response one level up? It would also allow more idiomatic error handling in this code, instead of the various early returns.